### PR TITLE
Migrated ctc_asr from keras2 to keras3

### PR DESCRIPTION
**Pull request by kharshith-k**
[Colab Notebook](https://colab.research.google.com/gist/kharshith-k/0f68c82611d35be40fd3eb97350c0c39/ctc_asr.ipynb)

### CODE REVIEW

File: examples/audio/ctc_asr.py [hunk 1]
```diff
@@ -2,11 +2,21 @@
...
 
+"""
+# Automatic Speech Recognition using CTC
+
+**Authors:** [Mohamed Reda Bouadjenek](https://rbouadjenek.github.io/) and [Ngoc Dung
+Huynh](https://www.linkedin.com/in/parkerhuynh/)<br>
+**Date created:** 2021/09/26<br>
+**Last modified:** 2026/01/22<br>
+**Description:** Training a CTC-based model for automatic speech recognition.
+"""
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: Remove this duplicate block

File: examples/audio/ctc_asr.py [hunk 2]
```diff
@@ -47,8 +57,9 @@
 - [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
 - [Speech recognition](https://en.wikipedia.org/wiki/Speech_recognition)
 - [Sequence Modeling With CTC](https://distill.pub/2017/ctc/)
-- [DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html)
-
+-
+[DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html)
+[DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html)
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: Remove duplicate line

File: examples/audio/ctc_asr.py [hunk 3]
```diff
@@ -245,20 +254,31 @@ def encode_single_sample(wav_file, label):
...
+    # Use TensorFlow's CTC loss (no backend-agnostic equivalent in Keras 3)
+    # blank_index=-1 means the blank label is the last class (output_dim + 1)
+    loss = tf.nn.ctc_loss(
+        labels=ops.cast(y_true, "int32"),
+        logits=y_pred,
+        label_length=label_length,
+        logit_length=input_length,
+        logits_time_major=False,
+        blank_index=-1,
+    )
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: We do have CTC loss in keras ops [CTC LOSS](https://keras.io/api/ops/nn/#ctcloss-function)

File: examples/audio/ctc_asr.py [hunk 4]
```diff
@@ -245,20 +254,31 @@ def encode_single_sample(wav_file, label):
...
     return loss
 
 
 """
 We now define our model. We will define a model similar to
 [DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html).
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py:274]: Remove duplicate

File: examples/audio/ctc_asr.py [hunk 5]
```diff
@@ -339,11 +359,22 @@ def build_model(input_dim, output_dim, rnn_layers=5, rnn_units=128):
...
     input_len = np.ones(pred.shape[0]) * pred.shape[1]
-    # Use greedy search. For complex tasks, you can use beam search
-    results = keras.backend.ctc_decode(pred, input_length=input_len, greedy=True)[0][0]
+
+    # Use TensorFlow's ctc_greedy_decoder (no backend-agnostic equivalent in Keras 3)
+    # Note: TF's decoder expects time-major format [time, batch, dim]
+    results, _ = tf.nn.ctc_greedy_decoder(
+        inputs=ops.transpose(pred, axes=[1, 0, 2]),
+        sequence_length=ops.cast(input_len, "int32"),
+    )
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: We have [Ctc decoder](https://keras.io/api/ops/nn/#ctcdecode-function) and use strategy as "greedy", if "greedy" is not working try with "beam search"

File: examples/audio/ctc_asr.py [hunk 6]
```diff
@@ -458,6 +487,9 @@ def on_epoch_end(self, epoch: int, logs=None):
 Example available on HuggingFace.
 | Trained Model | Demo |
 | :--: | :--: |
-| [![Generic badge](https://img.shields.io/badge/🤗%20Model-CTC%20ASR-black.svg)](https://huggingface.co/keras-io/ctc_asr) | [![Generic badge](https://img.shields.io/badge/🤗%20Spaces-CTC%20ASR-black.svg)](https://huggingface.co/spaces/keras-io/ctc_asr) |
-
+| [![Generic
+badge](https://img.shields.io/badge/🤗%20Model-CTC%20ASR-black.svg)](https://huggingface.co
+/keras-io/ctc_asr) | [![Generic
+badge](https://img.shields.io/badge/🤗%20Spaces-CTC%20ASR-black.svg)](https://huggingface.c
+o/spaces/keras-io/ctc_asr) |
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: Remove duplicate entry

File: examples/audio/ctc_asr.py [hunk 7]
```diff
@@ -2,9 +2,10 @@
 Title: Automatic Speech Recognition using CTC
 Authors: [Mohamed Reda Bouadjenek](https://rbouadjenek.github.io/) and [Ngoc Dung Huynh](https://www.linkedin.com/in/parkerhuynh/)
 Date created: 2021/09/26
-Last modified: 2021/09/26
+Last modified: 2026/01/27
 Description: Training a CTC-based model for automatic speech recognition.
 Accelerator: GPU
+Migrated By: [Harshith K](https://github.com/kharshith-k/)
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: Migrated By: --> Converted to Keras 3 by:

File: examples/audio/ctc_asr.py [hunk 8]
```diff
@@ -47,8 +48,8 @@
 - [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
 - [Speech recognition](https://en.wikipedia.org/wiki/Speech_recognition)
 - [Sequence Modeling With CTC](https://distill.pub/2017/ctc/)
-- [DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html)
-
+-
+[DeepSpeech2](https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/deepspeech2.html)
```
Comment (sachinprasadhs) [inline examples/audio/ctc_asr.py]: This should be in a single line
